account-kit/logging/src/index.ts

```diff
@@ -1,15 +1,54 @@
 import { createClientLogger } from "./client.js";
 import { createServerLogger } from "./server.js";
-import type { EventLogger, EventsSchema } from "./types";
+import type { EventLogger, EventsSchema, LoggerContext } from "./types";
 
 export type * from "./types.js";
 
-export function createLogger<
-  Schema extends EventsSchema
->(): EventLogger<Schema> {
-  if (typeof window === "undefined") {
-    return createServerLogger();
-  }
+export function createLogger<Schema extends EventsSchema = []>(
+  context: LoggerContext
+): EventLogger<Schema>;
 
-  return createClientLogger();
+export function createLogger(context: LoggerContext): EventLogger {
+  const innerLogger =
+    typeof window === "undefined"
+      ? createServerLogger(context)
+      : createClientLogger(context);
+
+  const logger: EventLogger = {
+    ...innerLogger,
+    profiled<TFunc extends (...args: any[]) => Promise<unknown> | unknown>(
+      name: string,
+      func: TFunc
+    ): TFunc {
+      return new Proxy(func, {
```

Comment:
This is fine, but the proxy feels unnecessary here. I think you could achieve the same thing by doing
```ts
return function(...args) {
  const start = Date.now();
  const result = func.apply(this, args);
  // ...
}
```

Reply:
ah yea, so that's exactly how I had it before and typescript wasn't happy about the typing of that lemme see if I can get it to work without the proxy

Reply:
The proxy is effectively just casting a bunch of stuff to `any` so if you had to you could do that with the nonproxy version and it wouldn't be any worse. But you might be able to do some clever typing to make it work, I dunno.

Reply:
I tried playing with it, one thing that may work would be to declare the types like this:
```ts
function profiled<TArgs extends any[], TRet>(func: (...args: TArgs) => TRet): (...args: TArgs) => TRet {
  return function (...args: TArgs): TRet {
    return func.apply(this, args);
  }
}
```
i.e. have separate type parameters for the parameters and return type. I'm not sure why it has trouble with just one param for the function type: it seems that TypeScript can't recognize that the types `TFunc` and ```ts
(...args: Parameters<TFunc>) => ReturnType<TFunc>
```
are the same.

Reply:
Oh you know, actually thinking about it more that makes sense. `TFunc`, despite extending a function, doesn't have to be _just_ a function: it could have other fields on it too. So returning a plain function doesn't necessarily make it the same type as `TFunc` because we'd be missing those extra fields.

Writing it in terms of parameters and return type makes it explicit that what you're getting back is _only_ the function piece of what was passed in, which practically speaking is all we need.

Reply:
lemme try this! thanks for looking into it

Reply:
Ok so single TFunc and using `Parameters<TFunc>` with `ReturnType<TFunc>` didn't work as typescript says they could be instantiated with different things.

what did work was making the generics on the `TArgs` and `TRet`, but I do still need to typecast in the promise case:

```
profiled<TArgs extends any[], TRet>(
      name: string,
      func: (...args: TArgs) => TRet
    ): (...args: TArgs) => TRet {
      return function (...args: TArgs): TRet {
        const start = Date.now();
        const result = func(...args);
        if (result instanceof Promise) {
          return result.then((res) => {
            innerLogger.trackEvent({
              name: "performance",
              data: {
                executionTimeMs: Date.now() - start,
                functionName: name,
              },
            });
            return res;
          }) as TRet; // Typecast is required here
        }

        innerLogger.trackEvent({
          name: "performance",
          data: {
            executionTimeMs: Date.now() - start,
            functionName: name,
          },
        });
        return result;
      };
    },
```

Reply:
If you did want to keep a single `TFunc` param, this declaration worked for me with the code you just pasted:
```ts
function profiled<TFunc extends (...args: unknown[]) => unknown>(
  name: string,
  func: TFunc,
): (...args: Parameters<TFunc>) => ReturnType<TFunc> {
```
that is, declaring the return type of `profiled` to be that derived type rather than `TFunc`. But that turns out more verbose than what you have so I like it as is.

Reply:
Also, I suspect that cast to `TRet` is needed for similar reasons to above: after the `instanceof Promise` check, we know it's a `Promise`, but it could also be more and have additional stuff on it (I've seen promises with `.cancel` methods on them, for example), whereas the thing we're returning is just a vanilla promise, so the type system can't assume that the return type is unchanged when you call `.then`.